library/entropy_poll.c

```diff
@@ -86,10 +92,16 @@ int mbedtls_platform_entropy_poll( void *data, unsigned char *output, size_t len
 
 /*
  * Test for Linux getrandom() support.
- * Since there is no wrapper in the libc yet, use the generic syscall wrapper
+ * When the C library is GNU libc and its version is greater than 2.25,
```

Comment:
Hold on, I don't think this is a good idea. You're testing the version of the Glibc headers that the library is compiled against. But you're relying on a runtime feature. On non-embedded platforms such as Linux with Glibc, it's very common to build a dynamically linked library on one machine and expect it to run on other machines that may not have the same libc version.

Compare the output of `make SHARED=1 && ldd -v library/libmbedcrypto.so` before and after this patch. After building on Ubuntu 20.04 with Glibc 2.31, before this patch:
```
        library/libmbedcrypto.so:
                libc.so.6 (GLIBC_2.7) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.14) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.4) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.2.5) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.3.4) => /lib/x86_64-linux-gnu/libc.so.6
```
After this patch:
```
                libc.so.6 (GLIBC_2.25) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.7) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.14) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.4) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.2.5) => /lib/x86_64-linux-gnu/libc.so.6
                libc.so.6 (GLIBC_2.3.4) => /lib/x86_64-linux-gnu/libc.so.6
```
Before the most recent dependency was on 2.14 [released](https://sourceware.org/glibc/wiki/Glibc%20Timeline) in June 2011. Now it's on 2.25 released in February 2017. That's quite a big jump. At a minimum, it should be announced in the changelog. But I'd prefer not to have this jump at all.

Reply:
ok thus if I understand you correctly the new line "libc.so.6 (GLIBC_2.25) => /lib/x86_64-linux-gnu/libc.so.6" is due to the call of getrandom() in mbedtls_platform_entropy_poll() added by this PR, right?

Reply:
Yes, this line appears when building 9c46a60e6c3edc35d6369afdf5252a5c66aef5b3 but not when building with 9c46a60e6c3edc35d6369afdf5252a5c66aef5b3~1, which confirms empirically that this patch caused it.

Reply:
[Patch to revert this PR and fix #3432 without breaking non-embedded Linux](https://github.com/ARMmbed/mbedtls/pull/3731)

Reply:
Does it mean that we can't use glibc's `getrandom` until we drop support for glibc < 2.25?
can we use some linking mechanism (for example wake symbol) to make use of glibc's `getrandom`?

Reply:
What's the point? What's the benefit of calling the glibc wrapper over the system call? The glibc wrapper adds a dependency: you need recent glibc on Linux, as opposed to just Linux. System calls on Linux are guaranteed to be stable.

We try to keep the makefiles simple so I don't think we'd accept glibc-specific linking code.

How would you do the fallback if the glibc `getrandom` isn't available? Wouldn't there need to be a runtime fallback to the system call anyway? If so why have both methods?